parsl/monitoring/file_monitoring.py

```diff
@@ -0,0 +1,488 @@
+    args: Any
+        The function arguments
+
+    Returns
+    -------
+    mp.Pool.AsyncResult
+    """
+    payload = dill.dumps((fun, args))
+    return monitor_pool.apply_async(run_dill_encoded, (payload,), callback=proc_callback)
+
+
+@typeguard.typechecked
+def monitor(task_id: int,
+            stop_event: Any,  # cannot be Event because of multiprocessing type weirdness.
+            done_event: Any,  # cannot be Event because of multiprocessing type weirdness.
+            patterns: List[Any],
+            callbacks: List[Callable],
+            sleep_dur: float,
+            work_dir: Optional[str] = None,
+            email: Optional[str] = None,
+            max_proc: int = 4) -> None:
+    """Function to monitor the file system for specific types of files and call a function when they are found.
+
+    This function runs in a periodic loop unitl it is told to stop. The time between loops is goverened by `sleep_dur`
+    seconds. Multiprocessing.Event objects are used to signal when this function should terminate. The general
+    workflow is as follows::
+                                        ------>callack (async)
+                                        |
+                     -----(found)-------+-------
+                     |                         |
+    Start----->scan for files---(none found)---+-->sleep --
+                    ^                                     |
+                    |                                     |
+                    ---------------------------------------
+
+    Any files that match any given pattern are tracked and only submitted to the callbacks once.
+
+    Parameters
+    ----------
+    task_id: int
+        The Parsl task id to be monitored.
+    stop_event: multiprocessing.Event
+        Signal for when to stop running the loop
+    done_event: nultiprocessing.Event
+        Signal used to indicate when this is actually completed, alowing for currently running callbacks to
+        complete nicely.
+    patterns: list
+        List of tuples containing a regex or glob type statement and a bool indicating whther it is regex (True) or not,
+        for finding the files.
+    callbacks: list
+        List of functions to call when files are found. This list must have either a single entry for all
+        `patterns` or have the same length as `patterns` (one callback for each pattern).
+    sleep_dur: float
+        The length of time to sleep between loops in seconds.
+    work_dir: str
+        The working directory for this process to run in. Default is ``None``, meaning the working directory is
+        inherited from the calling process.
+    email: str, optional
+        Email address to send callback results to
+    max_proc: int, optional
+        The maximum size of the multiprocessing Pool for the file processing. Default is 4, but the
+        value is adjusted to be ``min(max_proc, len(patterns))`` so as to not make the pool larger than
+        needed.
+    """
+    try:
```

Comment:
Does this code run on the submit side or on workers?
(aka does the user callback code run on the submit side or the worker side?)

I think it's running in a worker?

Reply:
Yes, on the worker, so perhaps the max_proc default should be reduced to 1 or 2 so as to not overload the worker. I should also add the caveat in the docs that any callbacks should be lightweight or risk slowing the worker.

Reply:
Another caveat to document, perhaps, is where side-effects of callbacks have their effect: the side-effects can only happen on the worker and cannot (for example) update a progress bar on the submitting process.

Reply:
I updated the documentation to make this clearer